R/staggregate.R

```diff
@@ -0,0 +1,508 @@
+
+  # Set key column in the climate data table
+  keycols = c("x", "y")
+  data.table::setkeyv(clim_dt, keycols)
+
+  # Keyed merge on the x/y column
+  merged_dt <- clim_dt[weights_dt, allow.cartesian = TRUE] #cols: x, y, date, value cols 1:k, poly_id, w_area, weight (if weights = T)
+
+  ## Multiply weights x climate value (all 1:k values); aggregate by month and polygon
+  ## -----------------------------------------------
+
+  # Multiply by secondary weights if weights = TRUE (already normalized by polygon area)
+  # Otherwise multiply by just area weights
+  if("weight" %in% names(merged_dt)){
+    merged_dt[, (list_names) := lapply(list_names, function(x) {get(x) * weight})]
+  } else {
+    merged_dt[, (list_names) := lapply(list_names, function(x) {get(x) * w_area})]
+  }
+
+  # Separate year, month, and day columns
+  merged_dt[, ':=' (year = substring(date, first = 1, last = 9),
+                    month = substring(date, first=10, last=17),
+                    day = substring(date, first=19))]
+
+  # Temporal Aggregation
+  if(agg_to == "year"){
+    sum_by_poly <- merged_dt[,  lapply(.SD, sum), by = .(poly_id, year),
+                             .SDcols = list_names]
+
+    ## Order columns
+    data.table::setcolorder(sum_by_poly, neworder = c('year', 'poly_id', list_names))
+  }
+
+  if(agg_to == "month"){
+    sum_by_poly <- merged_dt[,  lapply(.SD, sum), by = .(poly_id, year, month),
+                             .SDcols = list_names]
+
+    ## Order columns
+    data.table::setcolorder(sum_by_poly, neworder = c('year', 'month', 'poly_id', list_names))
+  }
+
+  if(agg_to == "day"){
+    sum_by_poly <- merged_dt[,  lapply(.SD, sum), by = .(poly_id, year, month, day),
+                             .SDcols = list_names]
+
+    ## Order columns
+    data.table::setcolorder(sum_by_poly, neworder = c('year', 'month', 'day', 'poly_id', list_names))
+  }
+
+
+
+  ## Return the sums by polygon
+  return(sum_by_poly)
+
+
+
+}
+
+#' @export
+staggregate_polynomial <- function(data, variable, daily_agg, overlay_weights, agg_to = "month", degree){
+
+  # Get climate data as a data table and aggregate to daily values
+  setup_list <- daily_aggregation(data, variable, daily_agg, overlay_weights)
+
+  clim_daily <- setup_list[[1]] # Pulls the daily aggregated raster brick
+  layer_names <- setup_list[[2]] # Pulls the saved layer names
+
+  # Polynomial transformation
+  poly_orders <- seq(1:degree) # Compute values from 1 to degree
+  list_length <- length(poly_orders) # How many lists are in the final object
+  list_names <- sapply(1:list_length, FUN=function(x){paste("order", poly_orders[x], sep="_")})
+
+  # For each daily layer, raise the value to degree, degree-1, degree-2 etc. until 1
+  r <- lapply(poly_orders, FUN=function(x){clim_daily ^ x})
+
+
+    ## Function: Set names of data table by month, change from wide to long format, rename based on polynomial orders
+  create_dt <- function(x){
```

Comment:
It seems like you could pull this `create_dt` function out of each `staggregate_*` function like you did with the `as.data.table.raster` function and place it at the top? We could probably also give it a better name if we do that so it's clearer what it's doing. Unless the function is slightly different depending on the aggregation type (I didn't check that)

Reply:
I think it remains the same. If that's the case, I will definitely extract it from each function, good catch.